Hardware instruction set support for crossgen2 (#33274)
dotnet/runtime: docs/design/coreclr/botr/vectors-and-intrinsics.md

Vectors and Hardware Intrinsics Support
===
---

# Introduction

> The CoreCLR runtime has support for several varieties of hardware intrinsics, and various ways to compile code which uses them. This support varies by what processor the logic is running on, and the code executed depends on how the RyuJIT was invoked. This document describes the various behaviors of intrinsics in the runtime, and concludes with documentation on what implications this has for developers working on the runtime and libraries portions of the runtime. 

Reviewer: I thought I'd read somewhere (not sure where) that we wanted to keep markdown line lengths below some reasonable length, eg 80 or 100?

There is decent tooling for this -- for instance VSCode markdown formatting extensions.

```suggestion
The CoreCLR runtime has support for several varieties of hardware intrinsics, and various ways to compile code which uses them. This support varies by target processor, and the code produced depends on how the jit compiler is invoked. This document describes the various behaviors of intrinsics in the runtime, and concludes with implications  for developers working on the runtime and libraries portions of the runtime. 
```

# Intrinsics apis
Most hardware intrinsics support is tied to the use of various Vector apis. There are 4 major api surfaces that are supported by the runtime

> - The fixed size float vectors. `Vector2`, `Vector3`, and `Vector4`. These vector types represent a struct of floats of various sizes. For type layout, ABI and, interop purposes they are represented in exactly the same way as a structure with an appropriate number of floats in it. Operations on these vector types are supported on all architectures and platforms, although some architectures may optimize various operations.

Reviewer: ```suggestion
- The fixed length float vectors. `Vector2`, `Vector3`, and `Vector4`. These vector types represent a struct of floats of various lengths. For type layout, ABI and, interop purposes they are represented in exactly the same way as a structure with an appropriate number of floats in it. Operations on these vector types are supported on all architectures and platforms, although some architectures may optimize various operations.
```

> - The variable length `Vector<T>`. This represents vector data of runtime-determined length. In any given process the length of a `Vector<T>` is the same in all methods, but this length may differ between various machines or environment variable settings read at startup of the process. The `T` type variable may be any of the primitive types, and allows use of integer or double data within a vector. The length and alignment of `Vector<T>` is unknown to the developer, and `Vector<T>` may not exist in any interop signature. Operations on these vector types are supported on all architectures and platforms, although some architectures may optimize various operations.

Reviewer 2: > The `T` type variable may be any of the primitive types

Maybe this could be reworded? It isn't allowed to be `char` or `bool` for example. Only the `8` basic integer types (signed and unsigned `int8`, `int16`, `int32`, and `int64`) and the `2` basic floating-point types (`float32` and `float64`)

Reviewer 2: > The length and alignment of `Vector<T>` is unknown to the developer, and `Vector<T>` may not exist in any interop signature

This isn't quite true, it is known at `Runtime` by checking `Vector<T>.Count` which specifies it is a struct of exactly `Count` `T` fields.

> - The variable sized `Vector<T>`. This represents vector data of indeterminate size. In any given process the size of a `Vector<T>` is the same in all methods, but this size may differ between various machines or environment variable settings read at startup of the process. The `T` type variable may be any of the primitive types, and allows use of integer or double data within a vector. The size and alignment of `Vector<T>` is unknown to the developer, and `Vector<T>` may not exist in any interop signature. Operations on these vector types are supported on all architectures and platforms, although some architectures may optimize various operations.

Reviewer: ```suggestion
- The variable length `Vector<T>`. This represents vector data of runtime-determined length. In any given process the length of a `Vector<T>` is the same in all methods, but this length may differ between various machines or environment variable settings read at startup of the process. The `T` type variable may be any of the primitive types, and allows use of integer or double data within a vector. The length and alignment of `Vector<T>` is unknown to the developer, and `Vector<T>` may not exist in any interop signature. Operations on these vector types are supported on all architectures and platforms, although some architectures may optimize various operations.
```

> - `Vector64<T>`, `Vector128<T>`, and `Vector256<T>` represent sized vectors that closely resemble the fixed size vectors available in C++. These structures can be used in any code that runs, but very few features are supported. These vector types define very few direct features, but instead most operations other than creation use the processor specific hardware intrinsics apis.

Reviewer: ```suggestion
- `Vector64<T>`, `Vector128<T>`, and `Vector256<T>` represent fixed-sized vectors that closely resemble the fixed- sized vectors available in C++. These structures can be used in any code that runs, but very few features are supported. These vector types define very few direct methods; most operations other than creation use the processor specific hardware intrinsics apis.
```

- Processor specific hardware intrinsics apis such as `System.Runtime.Intrinsics.X86.Ssse3`. These apis map directly to individual instructions or short instruction sequences that are specific to a particular hardware instruction. These apis are only useable on hardware that supports the particular instruction. See https://github.com/dotnet/designs/blob/master/accepted/2018/platform-intrinsics.md for the design of these.

# How to use intrinsics apis

There are 3 models for use of intrinsics apis.

1. Usage of `Vector2`, `Vector3`, `Vector4`, and `Vector<T>`. For these, its always safe to just use the types. The jit will generate code that is as optimal as it can for the logic, and will do so unconditionally.
2. Usage of `Vector64<T>`, `Vector128<T>`, and `Vector256<T>`. These types may be used unconditionally, but are only truly useful when also using the platform specific hardware intrinsics apis.
3. Usage of platform intrinsics apis. All usage of these apis should be wrapped in an `IsSupported` check of the appropriate kind. Then, within the `IsSupported` check, other platform specific apis may be used.

# Effect of usage of hardware intrinsics on how code is generated

> Hardware intrinsics have dramatic impacts on codegen, and the codegen of these hardware intrinsics is dependent on what knowledge of the target machine is available when the code is compiled.

Reviewer: ```suggestion
Hardware intrinsics have dramatic impacts on codegen, and the codegen of these hardware intrinsics is dependent on the ISA available for the target machine when the code is compiled.
```

> If the code is compiled at runtime by the JIT in a just-in-time manner, then the JIT will generate the best code it can for a given intrinsic usage when compiling tier 1 code, and generate functionally equivalent code when compiling tier 0 code. MethodImplOptions.AggressiveOptimization may be used to disable compilation of tier 0 code.

Reviewer: ```suggestion
If the code is compiled at runtime by the JIT in a just-in-time manner, then the JIT will generate the best code it can based on the current processor's ISA when compiling tier 1 code, and generate functionally equivalent code when compiling tier 0 code. MethodImplOptions.AggressiveOptimization may be used to bypass compilation of tier 0 code and always produce tier 1 code for the method.
```

> If the code is compiled at runtime by the JIT in a just-in-time manner, then the JIT will generate the best code it can based on the current processor's ISA when compiling tier 1 code, and generate functionally equivalent code when compiling tier 0 code. MethodImplOptions.AggressiveOptimization may be used to bypass compilation of tier 0 code and always produce tier 1 code for the method.

Reviewer 3: This tends to imply that the compiler will not recognize the intrinsics when compiling tier 0 code, but that's not the case. The recognition of intrinsics is independent of tier.

> For ahead of time compilation, the situation is far more complex. This is due to the following principles of how our ahead of time compilation model works.

Reviewer 3: I would suggest to use "ahead-of-time" to make it clear that this term is basically an adjective applied to `compilation`. In fact, I'd suggest to define it early, along with the acronym, and then use `AOT` in the rest of the document.

1. Ahead of time compilation must never under any circumstance change the semantic behavior of code except for changes in performance.
2. If ahead of time code is generated, it should be used unless there is an overriding reason to avoid using it.

> 3. It must be exceedingly difficult to misuse the ahead of time compilation tool to violate principle 1.

Reviewer 3: I would change "exceedingly" which implies "too much" to "extremely" or something.

> There are 2 different implementations of ahead of time compilation under development at this time. The crossgen1 model (which is currently supported on all platforms and architectures), and the crossgen2 model, which is under active development, and expected to replace crossgen1 at some point. Any developer wishing to use hardware intrinsics in the runtime or libraries should be aware of the crossgen1 model, which will hopefully be replaced with the crossgen2 model at some point. Thankfully, the restrictions implied on managed developers are strictly expected to be relaxed with crossgen2.

Reviewer: ```suggestion
There are 2 different implementations of ahead of time compilation under development at this time. The crossgen1 model (which is currently supported on all platforms and architectures), and the crossgen2 model, which is under active development. Any developer wishing to use hardware intrinsics in the runtime or libraries should be aware of the restrictions imposed by the crossgen1 model. Crossgen2, which we expect will replace crossgen1 at some point in the future, has strictly fewer restrictions.
```

## Crossgen1 model of hardware intrinsic usage

###Code written in System.Private.CoreLib.dll
#### Crossgen implementation rules
- Any code which uses `Vector<T>` will not be compiled ahead of time. (See code which throws a TypeLoadException using `IDS_EE_SIMD_NGEN_DISALLOWED`) 
- Code which uses Sse and Sse2 platform hardware intrinsics is always generated as it would be at jit time. 
- Code which uses Sse3, Ssse3, Sse41, Sse42, Popcnt, Pclmulqdq, and Lzcnt instruction sets will be generated, but the associated IsSupported check will be a runtime check. See `FilterNamedIntrinsicMethodAttribs` for details on how this is done.
- Code which uses other instruction sets will be generated as if the processor does not support that instruction set. (For instance, a usage of Avx2.IsSupported in CoreLib will generate native code where it unconditionally returns false, and then if and when tiered compilation occurs, the function may be rejitted and have code where the property returns true.)

> - Non-platform intrinsics which require more hardware support than the minimum supported hardware capability will not take advantage of that capability. In particular the code generated for `Vector2/3/4.Dot`, and `Math.Round`, and `MathF.Round`. See `FilterNamedIntrinsicMethodAttribs` for details. MethodImplOptions.AggressiveOptimization may be used to disable precompilation compilation of this sub-par code.

Reviewer 3: I think "precompilation compilation" should just be "precompilation"

#### Characteristics which result from rules
The rules here provide the following characteristics.
- Some platform specific hardware intrinsics can be used in CoreLib without encountering a startup time penalty

> - Use of platform specific hardware intrinsics causing runtime jit can be mitigated by careful choice of which intrinsics are used.

Reviewer 3: This is a bit confusing, perhaps a little elaboration would be useful.

> - Use of `Vector<T>` causes runtime jit and startup time concerns. Current analysis indicates this is acceptable, but it is a perennial concern for applications with tight startup time requirements.

Reviewer 3: after "concerns" add "because it is never precompiled".

        }
        else
            return FallbackPath(input);
    }

    private static int FallbackPath(uint)
    {
        const uint c1 = 0x_55555555u;
        const uint c2 = 0x_33333333u;
        const uint c3 = 0x_0F0F0F0Fu;
        const uint c4 = 0x_01010101u;

        value -= (value >> 1) & c1;
        value = (value & c2) + ((value >> 2) & c2);
        value = (((value + (value >> 4)) & c3) * c4) >> 24;

        return (int)value;
    }
}
```

UNACCEPTABLE code
```csharp
using System.Runtime.Intrinsics.X86;

public class BitOperations
{
    public static int PopCount(uint value)
    {
        if (Avx2.IsSupported)
            return UseAvx2(value);
        else
            return FallbackPath(input);
    }

    private static int FallbackPath(uint)
    {
        const uint c1 = 0x_55555555u;
        const uint c2 = 0x_33333333u;
        const uint c3 = 0x_0F0F0F0Fu;
        const uint c4 = 0x_01010101u;

        value -= (value >> 1) & c1;
        value = (value & c2) + ((value >> 2) & c2);
        value = (((value + (value >> 4)) & c3) * c4) >> 24;

        return (int)value;
    }

    private static int UseAvx2(uint value)
    {

>         // THIS IS A BUG!!!!!

Reviewer 2: Is it worth documenting that this is a bug because `PopCount` might have been rejitted to know `Avx2.IsSupported` but `UseAvx2` might have not?

Reviewer 2: Likewise, is this "ok" to do if `UseAvx2` is marked `AggressiveOptimization`?

Reviewer 3: I wouldn't think that this would be ok even if it is marked `AggressiveOptimization` - I don't think we should *ever* rely for correctness on the behavior of optimization.

Author: I agree with Carol here. At some point we may change the policy on `AggressiveOptimization`, and that shouldn't break the correctness of applications. We do parse another attribute, the Ssytem.Runtime.BypassReadyToRunAttribute, which can be used to simply declare that a method shouldn't be ready to runned. This attribute isn't actually in the BCL, the application must define it itself, but it can be used to safely forbid a method from being precompiled. I've updated the document, and added a comment that this pattern IS acceptable when applied to a subset of the x86 and x64 instruction sets, as we have special treatment for those instruction sets in the aot compiler when compiling CoreLib.

        Some series of Avx2 instructions that performs the popcount operation.
    }
}
```

### Code written in other assemblies (both first and third party)

#### Crossgen implementation rules

> - Any code which uses an intrinsic out of the `System.Runtime.Intrinsics.Arm` or `System.Runtime.Intrinsics.X86` namespace will not be compiled ahead of time. (See code which throws a TypeLoadException using `IDS_EE_HWINTRINSIC_NGEN_DISALLOWED`) 

Reviewer 3: Suggest "out of the" => "from the"

- Use of `Vector<T>` causes runtime jit and startup time concerns
- Ahead of time generated code which could take advantage of more advanced hardware support experiences a performance penalty until rejitted. (If a customer chooses to disable tiered compilation, then customer code may always run slowly). 

#### Code review rules for use of platform intrinsics
- Any use of a platform intrinsic in the codebase SHOULD be wrapped with a call to the associated IsSupported property. This wrapping may be done within the same function that uses the hardware intrinsic, but this is not required as long as the programmer can control all entrypoints to a function that uses the hardware intrinsic.
- If an application developer is highly concerned about startup performance, developers should avoid use of all platform specific hardware intrinsics on startup paths.

## Crossgen2 model of hardware intrinsic usage
There are 2 sets of instruction sets known to the compiler.
- The baseline instruction set which defaults to (Sse, Sse2), but may be adjusted via compiler option.
- The optimistic instruction set which defaults to (Sse3, Ssse3, Sse41, Sse42, Popcnt, Pclmulqdq, and Lzcnt).

> Code will be compiled using the optimistic instruction set to drive compilation, but any use of an instruction set beyond the baseline instruction set will be recorded, or any failure of use of an instruction set that would affect semantic behavior of code. If the baseline instruction set includes `Avx2` then the size and characteristics of of `Vector<T>` is known. Any other decisions about ABI may also be encoded. For instance, it is likely that the ABI of `Vector256<T>` will vary based on the presence/absence of `Avx` support.

Reviewer 3: I think the "will be recorded" belongs after the "or any failure" clause.

-  If an application developer is highly concerned about startup performance, developers should avoid use intrinsics beyond Sse42, or should use Crossgen with an updated baseline instruction set support.

### Crossgen2 ajustment to rules for System.Private.CoreLib.dll
Since System.Private.CoreLib.dll is known to be code reviewed with the code review rules as written above for crossgen1 with System.Private.CoreLib.dll, it is possible to relax rule "Code which attempts to use instruction sets outside of the optimistic set will generate code that will not be used on machines with support for the instruction set." What this will do is allow the generation of non-optimal code for these situations, but through the magic of code review, the generated logic will still work correctly.

# Mechanisms in the JIT to generate correct code to handle varied instruction set support

> The JIT receives flags which instruct it on what instruction sets are valid to use at, and the jit interface api s`notifyInstructionSetUsage(isa, bool supportBehaviorRequired)`. 

Reviewer: ```suggestion
The JIT receives flags which instruct it on what instruction sets are valid to use, and has access to a new jit interface api `notifyInstructionSetUsage(isa, bool supportBehaviorRequired)`. 
```

> The notifyInstructionSetUsage api is used to notify the AOT compiler infrastructure that the code may only execute if the runtime environment of the code is exactly the same as the boolean parameter indicates it should be. For instance, if `notifyInstructionSetUsage(Avx, false)` is used, then the code generated must not be used if the `Avx` instruction set is useable. Similarly `notifyInstructionSetUsage(Avx, true)` will indicate that the code may only be used if the `Avx` instruction set is available.

Reviewer: ```suggestion
The notifyInstructionSetUsage api is used to notify the AOT compiler infrastructure that the code may only execute if the runtime environment of the code is exactly the same as the boolean parameter indicates it should be. For instance, if `notifyInstructionSetUsage(Avx, false)` is used, then the code generated must not be used if the `Avx` instruction set is available. Similarly `notifyInstructionSetUsage(Avx, true)` will indicate that the code may only be used if the `Avx` instruction set is available.
```
